api/desecapi/models.py

```diff
@@ -464,78 +298,39 @@ class RRset(models.Model, mixins.SetterMixin):
     class Meta:
         unique_together = (("domain", "subname", "type"),)
 
-    def __init__(self, *args, **kwargs):
-        self._dirties = set()
-        super().__init__(*args, **kwargs)
-
-    def setter_domain(self, val):
-        if val != self.domain:
-            self._dirties.add('domain')
-
-        return val
-
-    def setter_subname(self, val):
-        # On PUT, RRsetSerializer sends None, denoting the unchanged value
-        if val is None:
-            return self.subname
-
-        if val != self.subname:
-            self._dirties.add('subname')
-
-        return val
-
-    def setter_type(self, val):
-        if val != self.type:
-            self._dirties.add('type')
-
-        return val
+    @property
+    def name(self):
+        return '.'.join(filter(None, [self.subname, self.domain.name])) + '.'
 
-    def setter_ttl(self, val):
-        if val != self.ttl:
-            self._dirties.add('ttl')
+    def save(self, *args, **kwargs):
+        self.updated = timezone.now()
+        super().save(*args, **kwargs)
 
-        return val
+    def set_all_record_contents(self, record_contents):
+        """
+        Updates this RR set's resource records, discarding any old values.
 
-    def clean(self):
-        errors = {}
-        for field in (self._dirties & {'domain', 'subname', 'type'}):
-            errors[field] = ValidationError(
-                'You cannot change the `%s` field.' % field)
+        To do so, two large select queries and one query per changed (added or removed) resource record are needed.
 
-        if errors:
-            raise ValidationError(errors)
+        Changes are saved to the database immediately.
 
-    def get_dirties(self):
-        return self._dirties
+        :param record_contents: set of strings
+        """
+        # Remove RRs that we didn't see in the new list
+        removed_rrs = self.records.exclude(content__in=record_contents)  # one SELECT
```

Comment:
You seem to need all records anyways, i.e. the union of the two mutually SELECTs will give `self.records.all()`. It may be nicer to iterate over that, and figure the `in record_contents` logic out on the application side?

This would become even more attractive with a streaming iterator, https://docs.djangoproject.com/en/2.2/ref/models/querysets/#iterator. Unfortunately, there's no benefit with MySQL (although it should work). If we use this now, we'll probably have a small gain without extra work in case we should switch to Postgres. (On the other hand, the streaming iterator does not have caching. But I suppose we would not expect this query to be cached, would we?)

Reply:
Yes, I'm doing two requests here to make the code easier to read. My attempts to save the second query all resulted in messy code, but I'm open to any neat suggestions to improve the efficiency here.

We should also consider using bulk operations here, notifying the change tracker by some other means.

Reply:
jut, probably also an efficiency issue we can deal with later